src/charm.py

```diff
@@ -35,85 +33,194 @@ def __init__(self, *args: typing.Any):
+            return False
+
+    def _configure_target_proxy(self) -> bool:
+        """Configure aproxy snap with proxy settings.
+
+        Returns:
+            True if target proxy is successfully configured, False otherwise.
+        """
+        if not self._target_proxy:
+            self.unit.status = ops.BlockedStatus("Missing target proxy address in config.")
+            return False
+
+        if not self._is_proxy_reachable(self._target_proxy, 3128):
+            logger.warning("Proxy is not reachable at %s:3128", self._target_proxy)
+            self.unit.status = ops.BlockedStatus(
+                f"Target proxy is unreachable at {self._target_proxy}:3128."
+            )
+            return False
+
+        try:
+            # nosec B404,B603,B607: calling trusted system binary with predefined args
+            subprocess.run(
+                ["snap", "set", "aproxy", f"proxy={self._target_proxy}:3128"],  # nosec
+                check=True,
+            )
+            logger.info("Configured aproxy snap with target proxy=%s:3128.", self._target_proxy)
+        except subprocess.CalledProcessError as e:
+            logger.error("Failed to configure aproxy snap: %s", e)
+            self.unit.status = ops.BlockedStatus("Failed to configure aproxy snap.")
+            return False
+        return True
+
+    def _format_ports(self, ports: str) -> str:
+        """Format a comma-separated list of ports into nftables port set syntax.
 
-    def _on_config_changed(self, event: ops.ConfigChangedEvent) -> None:
-        """Handle changed configuration.
+        Args:
+            ports: Comma-separated string of port numbers.
+        """
+        if ports.strip().upper() == "ALL":
+            return "0-65535"
+        return ", ".join(port.strip() for port in ports.split(",") if port.strip())
 
-        Change this example to suit your needs. If you don't need to handle config, you can remove
-        this method.
+    def _apply_nftables_rules(self) -> bool:
+        """Apply nftables rules for transparent proxy interception.
 
-        Learn more about config at
-        https://canonical-charmcraft.readthedocs-hosted.com/stable/reference/files/config-yaml-file/
+        - Redirect outbound traffic on configured intercept_ports to aproxy (127.0.0.1:8443).
+        - Exclude private and loopback ranges.
+        - Drop inbound traffic to aproxy listener to prevent reflection attacks.
 
-        Args:
-            event: event triggering the handler.
+        Returns:
+            True if rules were successfully applied, False otherwise.
+        """
+        no_proxy_list = [ip.strip() for ip in self._no_proxy.split(",") if ip.strip()]
+        no_proxy_clause = (
+            f"ip daddr {{ {', '.join(no_proxy_list)} }} return" if no_proxy_list else ""
+        )
+        ports_clause = self._format_ports(self._intercept_ports)
+
+        rules = f"""
+        table ip aproxy
```

Comment:
Consider copying the new nftables format from the GitHub runner project, which is much cleaner:


https://github.com/canonical/github-runner-operator/blob/77d1ed40a85e632092c8dcf0309b4959d8ce04ee/github-runner-manager/src/github_runner_manager/templates/openstack-userdata.sh.j2#L18-L34